ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java

```diff
@@ -1605,86 +1640,134 @@ private void updateIdealStateWithNewPartitionsForFullAuto(String dcName,
-      currentInstances.addAll(lastEntry.getValue());
-      info("[{}] Already has resources, will try to use last resource {}, current size: {}, target size {}",
-          dcName.toUpperCase(), startResourceId, currentInstances.size(), maxInstancesInOneResourceForFullAuto);
+      startResourceId = lastEntry.getKey().intValue() + 1;
     }
-    int curResourceId = startResourceId;
+    // Break these smaller instance groups into bigger new groups
+    List<Set<String>> newGroups = new ArrayList<>();
+    Set<String> currentInstances = new HashSet<>();
     while (instanceGroupIterator.hasNext()) {
       Set<String> group = instanceGroupIterator.next();
       if (currentInstances.size() + group.size() > maxInstancesInOneResourceForFullAuto) {
-        resourceIdToInstances.put(curResourceId, new HashSet<>(currentInstances));
-        curResourceId++;
+        newGroups.add(new HashSet<>(currentInstances));
         currentInstances.clear();
       }
       currentInstances.addAll(group);
     }
     if (currentInstances.size() != 0) {
-      resourceIdToInstances.put(curResourceId, currentInstances);
+      newGroups.add(currentInstances);
     }
 
-    for (; startResourceId <= curResourceId; startResourceId++) {
-      String resourceName = Integer.toString(startResourceId);
-      List<Map.Entry<String, Set<String>>> partitionsUnderResource = resourceIdToInstances.get(startResourceId)
-          .stream()
-          .map(in -> instanceToPartitions.get(in))
-          .filter(Objects::nonNull) // instances in last resource id might not have partitions in this map
-          .flatMap(ps -> ps.stream())
-          .collect(Collectors.toSet())
-          .stream()
-          .map(p -> new AbstractMap.SimpleEntry<>(p, partitionsToInstancesInDc.get(p)))
-          .collect(Collectors.toList());
-      if (startResourceId == maxResource) {
-        if (!partitionsUnderResource.isEmpty()) {
-          // Update ideal state for this resource
-          IdealState idealState = resourceIdToIdealState.get(startResourceId);
-          int sizeBefore = idealState.getPartitionSet().size();
-          updateIdealStatePartitions(idealState, partitionsUnderResource);
-          if (!dryRun) {
-            dcAdmin.setResourceIdealState(clusterName, resourceName, idealState);
-            info("[{}] Update resource {} in datacenter {} from {} partitions to {} partitions", dcName.toUpperCase(),
-                resourceName, dcName, sizeBefore, partitionsUnderResource.size());
-          } else {
-            info("[{}] Under DryRun mode, update resource {} in datacenter {} from {} partitions to {} partitions",
-                dcName.toUpperCase(), resourceName, dcName, sizeBefore, partitionsUnderResource.size());
-          }
-          resourcesUpdated.incrementAndGet();
-        } else {
-          info("[{}] No changes made to resource {}, skip", dcName.toUpperCase(), resourceName);
+    for (Set<String> instances : newGroups) {
```

Comment:
What's the "group" concept in the Helix concept?

Reply:
Groups means a set of instances that share partitions with each other. Instances in one group should never share partitions with any instances in other groups.